pkg/lease/manager.go

```diff
@@ -0,0 +1,235 @@
+package lease
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	log "github.com/sirupsen/logrus"
+
+	coordv1 "k8s.io/api/coordination/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apitypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/pointer"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Manager interface {
+	//RequestLease will create a lease with leaseDuration if it does not exist or extend existing lease duration to leaseDuration.
+	//It'll return an error in case it can't do either (for example if the lease is already taken).
+	RequestLease(ctx context.Context, obj client.Object, leaseDuration time.Duration) error
+	//InvalidateLease will release the lease.
+	InvalidateLease(ctx context.Context, obj client.Object) error
+	//GetLease will try to fetch a lease.
+	//It'll return an error in case it can't (for example if the lease does not exist or is already taken).
+	GetLease(ctx context.Context, obj client.Object) (*coordv1.Lease, error)
+}
+
+type manager struct {
+	client.Client
+	holderIdentity string
+	namespace      string
+	log            logr.Logger
+}
+
+func (l *manager) RequestLease(ctx context.Context, obj client.Object, leaseDuration time.Duration) error {
```

Comment:
What is the purspose of this pattern,letting the public method just call a private method with the same signature? 🤔

Reply:
> separation between the API (public method) which needs to follow up the interface signature 

But the interface is the API, not? While the methods here are public, their receiver (the `manager` struct) isn't. So in the end we are private here already.

> it better protects the signature of the API.

 I don't see how this protects the API. As long as the interface doesn't change, the API is safe.


> I still see value in this separation [...] conceptually

I can't remember seeing this pattern before, is this best practice? Maybe I'm just missing something 🤔

Reply:
>  it's fairly easy to refactor RequestLease so it'll change the API as well - most IDEs will do it for you.

I still doubt the value over the confusion, but this is a point at least. So I'm fine with keeping it.